packages/fx-core/src/plugins/solution/fx-solution/arm.ts

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import {
+  AzureSolutionSettings,
+  SolutionContext,
+  Plugin,
+  Result,
+  ok,
+  FxError,
+} from "@microsoft/teamsfx-api";
+import { ScaffoldArmTemplateResult, ArmResourcePlugin } from "../../../common/armInterface";
+import { getActivatedResourcePlugins } from "./ResourcePluginContainer";
+import { getPluginContext } from "./util";
+import { format } from "util";
+import { compileHandlebarsTemplateString, getStrings } from "../../../common";
+import path from "path";
+import * as fs from "fs-extra";
+
+const baseFolder = "./infra/azure";
+const templateFolder = "templates";
+const parameterFolder = "parameters";
+const bicepOrchestrationFileName = "main.bicep";
+const parameterTemplateFileName = "parameter.template.json";
+const solutionLevelParameters = `param resourceBaseName string\n`;
+
+// Get ARM template content from each resource plugin and output to project folder
+export async function generateArmTemplate(ctx: SolutionContext): Promise<Result<any, FxError>> {
+  const azureSolutionSettings = ctx.projectSettings?.solutionSettings as AzureSolutionSettings;
+  const plugins = getActivatedResourcePlugins(azureSolutionSettings); // This function ensures return result won't be empty
+
+  const bicepOrchestrationTemplate = new BicepOrchestrationTemplate(plugins.map((p) => p.name));
+  const moduleFiles = new Map<string, string>();
+
+  // Get bicep content from each resource plugin
+  for (const plugin of plugins) {
+    const pluginWithArm = plugin as Plugin & ArmResourcePlugin; // Temporary solution before adding it to teamsfx-api
+    if (pluginWithArm.scaffoldArmTemplate) {
+      // find method using method name
+      const pluginContext = getPluginContext(ctx, pluginWithArm.name);
+      const result = (await pluginWithArm.scaffoldArmTemplate(pluginContext)) as Result<
+        ScaffoldArmTemplateResult,
+        FxError
+      >;
+      if (result.isOk()) {
+        bicepOrchestrationTemplate.addTemplate(pluginWithArm.name, result.value);
+        if (result.value.Modules) {
+          for (const module of Object.entries(result.value.Modules)) {
+            const moduleFileName = module[0];
+            const moduleFileContent = module[1].Content;
+            moduleFiles.set(generateBicepModuleFilePath(moduleFileName), moduleFileContent);
+          }
+        }
+      } else {
+        const msg = format(getStrings().solution.ProvisionFailNotice, ctx.projectSettings?.appName);
```

Comment:
Do we need send telemetry for the failure

Reply:
Thanks, we will review the telemetry strategy and add them later. Have created one task to track this.